cupyx/texture/_affine.py

```diff
@@ -0,0 +1,159 @@
+
+    if read_mode == 'element_type':
+        read_mode = runtime.cudaReadModeElementType
+    elif read_mode == 'normalized_float':
+        read_mode = runtime.cudaReadModeNormalizedFloat
+    else:
+        raise ValueError(
+            'Unsupported read mode '
+            '(supported: element_type, normalized_float)')
+
+    texture_fmt = texture.ChannelFormatDescriptor(
+        data.itemsize * 8, 0, 0, 0, fmt_kind)
+    # print(f'{data.shape=}, {data.itemsize=}, {data.dtype=}')
+    # CUDAArray: last dimension=fastest changing dimension
+    array = texture.CUDAarray(texture_fmt, *data.shape[::-1])
+    res_desc = texture.ResourceDescriptor(
+        runtime.cudaResourceTypeArray, cuArr=array)
+    tex_desc = texture.TextureDescriptor(
+        (address_mode, ) * data.ndim, filter_mode, read_mode)
+    tex_obj = texture.TextureObject(res_desc, tex_desc)
+    array.copy_from(data)
+
+    return tex_obj
+
+
+@cupy._util.memoize(for_each_device=True)
+def _get_affine_transform_kernel(ndim: int):
+
+    if ndim == 2:
+        return cupy.ElementwiseKernel(
+            'U texObj, raw float32 m, uint64 width',
+            'T transformed_image',
+            '''
+            float3 pixel = make_float3(
+                (float)(i / width) + .5f,
+                (float)(i % width) + .5f,
+                1.0f
+            );
+
+            float x = dot(pixel, make_float3(m[0],  m[1],  m[2]));
+            float y = dot(pixel, make_float3(m[3],  m[4],  m[5]));
+
+            transformed_image = tex2D<T>(texObj, y, x);
+            ''',
+            'affine_transformation_2d',
+            preamble='''
+            inline __host__ __device__ float dot(float3 a, float3 b)
+            {
+                return a.x * b.x + a.y * b.y + a.z * b.z;
+            }
+            ''')
+
+    elif ndim == 3:
+        return cupy.ElementwiseKernel(
+            'U texObj, raw float32 m, uint64 height, uint64 width',
+            'T transformed_volume',
+            '''
+            float4 voxel = make_float4(
+                (float)(i / (width * height)) + .5f,
+                (float)((i % (width * height)) / width) + .5f,
+                (float)((i % (width * height)) % width) + .5f,
+                1.0f
+            );
+
+            float x = dot(voxel, make_float4(m[0],  m[1],  m[2],  m[3]));
+            float y = dot(voxel, make_float4(m[4],  m[5],  m[6],  m[7]));
+            float z = dot(voxel, make_float4(m[8],  m[9],  m[10], m[11]));
+
+            transformed_volume = tex3D<T>(texObj, z, y, x);
+            ''',
+            'affine_transformation_3d',
+            preamble='''
+            inline __host__ __device__ float dot(float4 a, float4 b)
+            {
+                return a.x * b.x + a.y * b.y + a.z * b.z + a.w * b.w;
+            }
+            ''')
+    else:
+        raise ValueError(
+            'Texture memory affine transformation is defined '
+            'only for single channel 2D and 3D arrays')
```

Comment:
Forgive my ignorance of `ndimage` APIs, I am just wondering if any of the `ndimage` APIs supports batch processing (say transforming a stack of images at once). Looks to me the current implementation only supports single-image processing.

Reply:
The SciPy ndimage APIs don't support batch processing (or vector-valued pixels). One has to call the function separately on each channel of a color image, for example. 

This does result in some inefficiency due to redundant computation since the interpolation weights should be the same for all channels!  I would be interested in having better support for `float3` (e.g. RGB images) and `float4` (r.g. RGBA) cases at a minimum for the existing functions in `cupyx.scipy.ndimage`. I don't recall offhand if we have any tests or examples using these types with `ElementwiseKernel`s?

Reply:
@the-lay @grlee77  Looks like we're not going to support multi-channel texture (which I am totally fine). But, I am still curious about the use case of transforming a stack/batch of images. I think this is possible with `ndimage` API:
```python
>>> import cupy as cp
>>> import cupyx.scipy.ndimage as img
>>>
>>> a = cp.random.random((4, 16, 16))  # 4 images of size 16*16
>>> b = cp.random.random((3, 3))
>>> out = img.affine_transform(a, b)
>>> out.shape
(4, 16, 16)
``` 
but perhaps to make this work with texture memory we need to support [layered texture](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#layered-textures)? If it's the case, I am happy to play with it and see if I can create a PR once this one is in.

Reply:
Actually I will try to implement support for up to 4 channel in the next commit, that should be doable, as long as the data type is fixed to float32.

I didn't know about layered textures, looks interesting! It seems they support only 1D and 2D, but as many channels (layers) as needed.

Reply:
Ah yes, layered texture is only for 1D & 2D. But I think transforming a stack of 2D images is a common task. I will try to put together some code and test the performance when I have time.